subprojects/files/src/main/java/org/gradle/api/internal/file/AbstractFileCollection.java

```diff
@@ -214,12 +206,24 @@ public TaskDependency getBuildDependencies() {
 
             @Override
             public boolean contains(File file) {
-                return AbstractFileCollection.this.contains(file) && predicate.apply(file);
+                return AbstractFileCollection.this.contains(file) && filterSpec.isSatisfiedBy(file);
             }
 
             @Override
             public Iterator<File> iterator() {
-                return Iterators.filter(AbstractFileCollection.this.iterator(), predicate);
```

Comment:
Why can't we use Java 8 for implementing this method?

Reply:
We can. I was aiming for minimal changes in this PR though.

Reply:
Why don't we continue to use `Iterators.filter`?

Reply:
I first started to rewrite everything not to use `Predicate`, but then I decided to keep changes to a minimum, and reverted most of them. Here I kept my original changes because I think it's a genuine improvement, as we stopped wrapping a `Spec` in a `Predicate`. I can change it back to `Iterable` if you feel strongly.

BTW, I expect that once Java 8 syntax is enabled for all this affected code we'll drop `Iterables` etc. and start using streams, and thus most of the code changed here will be rewritten again.